### ISSUE TITLE
How to exclude a certain package from `home.packages`?

### ISSUE THREAD

**Issue by rennsax** [nix-community/home-manager#5658]
### Description

I do not want to include `pkgs.zsh` in `home.packages` since it may conflict with the one that is installed at system-width (I use Nix on many distros). But I find it's hard-coded here:

https://github.com/nix-community/home-manager/blob/7560dc942a6fbd37ebd1310b3dbda513de2d4b82/modules/programs/zsh.nix#L564-L566

And `programs.zsh.package` is not `nullable` (https://github.com/NixOS/nixpkgs/blob/27ab0078d8c9578583f6891ca506fb8c1e13b37c/lib/options.nix#L192). Is there any workaround to exclude `pkgs.zsh` from `home.packages`? I think I can copy my own `zsh.nix` and just modify that line. Is there any simpler trick?

**Comment by mainrs** [comment]
I had the same issue yesterday. Take a look at my config:

https://github.com/mainrs/nixos/blob/refactor/homes/x86_64-linux/me/default.nix#L1-L27

**Comment by rennsax** [comment]
@mainrs Thanks for your advice! This is a good resolution, but in my case, overriding `programs.<program>.package` is inappropriate because the module may use `cfg.package` to define some of the configurations, just like:

https://github.com/nix-community/home-manager/blob/304a011325b7ac7b8c9950333cd215a7aa146b0e/modules/programs/zsh.nix#L587

However, I find my use case really strange. Home-manager uses `cfg.package` to generate zsh init files, so installing `cfg.package` is almost necessary.

**Comment by ryan4yin** [comment]
## TLDR

You need to customize the zshrc generated by home-manager by:

```nix
{pkgs}: {
  programs.zsh = {
    package = pkgs.emptyDirectory;
    # NOTE: this won't work on NixOS!
    initExtraBeforeCompInit = ''
      HELPDIR="/usr/share/zsh/$ZSH_VERSION/help"
    '';
  };
}
```

Haven't tested it, hope it works for you~

## Details

A common solution is add something like `programs.zsh.package = pkgs.emptyDirectory;` into your home-manager's config, but since home-manager's zsh module dose reference `/share/zsh/$ZSH_VERSION/help` in zsh's nix package, this will make `HELPDIR` unable to work.

I tried a wrokaround like this:

```nix
{pkgs, ...}: {
  programs.zsh.package = pkgs.stdenv.mkDerivation {
    name = "myzsh";
    phases = ["installPhase"];
    installPhase = ''
      mkdir -p $out
      cp -r ${pkgs.zsh}/share $out/share
    '';
  };
}
```

But it did not resolve the issue, since zsh contains its version in the helpdir's path:

1. my custom zsh contains only 5.9's share files:
    ```bash
    › ls -al /nix/store/l10f6l44xxxjhcv0c3qvqgq49sji6ljd-myzsh/share/zsh/
    total 0
    dr-xr-xr-x 3 root wheel  96 Jan  1  1970 .
    dr-xr-xr-x 3 root wheel  96 Jan  1  1970 ..
    dr-xr-xr-x 5 root wheel 160 Jan  1  1970 5.9
    
    › ls -al /nix/store/l10f6l44xxxjhcv0c3qvqgq49sji6ljd-myzsh/share/zsh/5.9
    total 0
    dr-xr-xr-x    5 root wheel   160 Jan  1  1970 .
    dr-xr-xr-x    3 root wheel    96 Jan  1  1970 ..
    dr-xr-xr-x 1206 root wheel 38592 Jan  1  1970 functions
    dr-xr-xr-x  109 root wheel  3488 Jan  1  1970 help
    dr-xr-xr-x    3 root wheel    96 Jan  1  1970 scripts
    ```
2. The zshrc generated by home-manager will use the current version of zsh to splice the HELPDIR:
    ```bash
    › cat ~/.zshrc | head -n 20
    
    typeset -U path cdpath fpath manpath
    
    for profile in ${(z)NIX_PROFILES}; do
      fpath+=($profile/share/zsh/site-functions $profile/share/zsh/$ZSH_VERSION/functions $profile/share/zsh/vendor-completions)
    done
    
    HELPDIR="/nix/store/dzma3zw4zx0d0xqwvbh5dq69i2sz0x7v-zsh-5.9/share/zsh/$ZSH_VERSION/help"
    
    
    # Oh-My-Zsh/Prezto calls compinit during initialization,
    # calling it twice causes slight start up slowdown
    # as all $fpath entries will be traversed again.
    autoload -U compinit && compinit
    ```

So by this way, when the zsh on your host is not the same as the zsh package you used in home-manager, `HELPDIR` will not work...

To fix the issue, we have to overwrite `HELPDIR` by adding our custom init scripts into `~/.zshrc`, here is a workaround on non-NixOS:

```nix
{pkgs}: {
  programs.zsh = {
    package = pkgs.emptyDirectory;
    # NOTE: this won't work on NixOS!
    initExtraBeforeCompInit = ''
      HELPDIR="/usr/share/zsh/$ZSH_VERSION/help"
    '';
  };
}
```

As for NixOS, I would recommend manage your zsh via home-manager directly to void all those tricks.

**Comment by rennsax** [comment]
Thanks for advice from both you two!
